librz/core/cbin.c

```diff
@@ -4961,42 +4959,112 @@ RZ_API bool rz_core_bin_memory_print(RzCore *core, RzBinFile *bf, RzCmdStateOutp
 	return true;
 }
 
-RZ_API bool rz_core_bin_resources_print(RzCore *core, RzBinFile *bf, RzCmdStateOutput *state) {
+static void bin_resources_print_standard(RzCore *core, RzList *hashes, RzBinResource *resource) {
+	char humansz[8];
+	rz_num_units(humansz, sizeof(humansz), resource->size);
+	rz_cons_printf("Resource %zd\n", resource->index);
+	rz_cons_printf("  name: %s\n", resource->name);
+	rz_cons_printf("  timestamp: %s\n", resource->time);
+	rz_cons_printf("  vaddr: 0x%08" PFMT64x "\n", resource->vaddr);
+	rz_cons_printf("  size: %s\n", humansz);
+	rz_cons_printf("  type: %s\n", resource->type);
+	rz_cons_printf("  language: %s\n", resource->language);
+	if (hashes && resource->size > 0) {
+		HtPP *digests = rz_core_bin_create_digests(core, resource->vaddr, resource->size, hashes);
+		if (!digests) {
+			return;
+		}
+		RzListIter *it = NULL;
+		char *hash = NULL;
+		bool found = false;
+		rz_list_foreach (hashes, it, hash) {
+			char *digest = ht_pp_find(digests, hash, &found);
+			if (found) {
+				rz_cons_printf("  %s: %s\n", hash, digest);
+			}
+		}
+		ht_pp_free(digests);
+	}
+}
+
+static void bin_resources_print_table(RzCore *core, RzCmdStateOutput *state, RzList *hashes, RzBinResource *resource) {
+	rz_table_add_rowf(state->d.t, "dssXxss", resource->index, resource->name,
+		resource->type, resource->vaddr, resource->size, resource->language, resource->time);
+	if (hashes && resource->size > 0) {
+		HtPP *digests = rz_core_bin_create_digests(core, resource->vaddr, resource->size, hashes);
+		if (!digests) {
+			return;
+		}
+		RzListIter *it;
+		char *hash;
+		bool found = false;
+		rz_list_foreach (hashes, it, hash) {
+			char *digest = ht_pp_find(digests, hash, &found);
+			if (found && state->d.t) {
+				rz_table_add_row_columnsf(state->d.t, "s", digest);
+			}
+		}
+		ht_pp_free(digests);
+	}
+}
+
+static void bin_resources_print_json(RzCore *core, RzCmdStateOutput *state, RzList *hashes, RzBinResource *resource) {
+	pj_o(state->d.pj);
+	pj_ks(state->d.pj, "name", resource->name);
+	pj_ki(state->d.pj, "index", resource->index);
+	pj_ks(state->d.pj, "type", resource->type);
+	pj_kn(state->d.pj, "vaddr", resource->vaddr);
+	pj_ki(state->d.pj, "size", resource->size);
+	pj_ks(state->d.pj, "lang", resource->language);
+	pj_ks(state->d.pj, "timestamp", resource->time);
+	if (hashes && resource->size > 0) {
+		HtPP *digests = rz_core_bin_create_digests(core, resource->vaddr, resource->size, hashes);
+		if (!digests) {
+			goto end;
+		}
+		RzListIter *it;
+		char *hash;
+		bool found = false;
+		rz_list_foreach (hashes, it, hash) {
+			char *digest = ht_pp_find(digests, hash, &found);
+			if (found && state->d.pj) {
+				pj_ks(state->d.pj, hash, digest);
+			}
+		}
+		ht_pp_free(digests);
+	}
+end:
+	pj_end(state->d.pj);
+}
+
+RZ_API bool rz_core_bin_resources_print(RzCore *core, RzBinFile *bf, RzCmdStateOutput *state, RzList *hashes) {
```

Comment:
Allow `hashes` to be NULL, add `RZ_NULLABLE`, also call `rz_list_foreach` only when hashes are not NULL.

Reply:
actually
```c
#define rz_list_foreach(list, it, pos) \
	if (list) \
		for (it = list->head; it && (pos = it->data, 1); it = it->n)
```